### PR TITLE
Add the flash address of image B and fix the flash driver bug

### PR DESCRIPTION
**Pull request by ghost**
The change involves implementation of JIRA FW-194, and a bug fix to the flash driver. Without the fix, the flash will become not writable and the FW-194 can not be tested.

### CODE REVIEW

File: common/cmd_bootm.c [hunk 1]
```diff
@@ -159,6 +159,7 @@ extern void lynxkdi_boot( image_header_t * );
 
 image_header_t header;
 ulong load_addr =  CFG_LOAD_ADDR;		/* Default Load Address */
+ulong sys_bootflag =  0x5A5A5A5A;		/* Default Load Address */
```
Comment (truongn1) [inline common/cmd_bootm.c]: better to have a definition elsewhere: for instant #define BOOT_FROM_A 0x5A5A5A5A

File: common/cmd_nvedit.c [hunk 2]
```diff
@@ -203,6 +204,16 @@ int _do_setenv (int flag, int argc, char *argv[])
 		}
 #endif
 
+		if (strcmp(argv[1],"bootflag") == 0) {
+			ulong flag = simple_strtoul(argv[2], NULL, 16);
+			if (0x5A5A5A5A != flag && 0xA5A5A5A5 != flag) {
+				printf("wrong flag, the bootflag can only be either 0x5A5A5A5A (for #img A) or 0xA5A5A5A5 (for #img B)\n");
+				return 1;
```
Comment (truongn1) [inline common/cmd_nvedit.c:211]: What if we got a wrong boot flag, would device will boot? I think we should try to boot from A (or B) if something wrong with the flag

File: net/httpd.c [hunk 3]
```diff
@@ -82,10 +84,23 @@ int do_http_upgrade(const ulong size, const int upgrade_type){
         OLED_Clear_page(6);
         OLED_ShowString(0,2,"DownLoad:100%");
 #endif
-        ret = raspi_erase_write((unsigned char*)(WEBFAILSAFE_UPLOAD_RAM_ADDRESS),WEBFAILSAFE_UPLOAD_KERNEL_ADDRESS,size);
+
+		unsigned int upgrade_flash_offset;
+		
+		if (0x5A5A5A5A == sys_bootflag) {
+			upgrade_flash_offset = 0x50000;
```
Comment (truongn1) [inline net/httpd.c]: We should change all hardcoded value here to macro define

File: common/cmd_nvedit.c [hunk 4]
```diff
@@ -203,6 +204,16 @@ int _do_setenv (int flag, int argc, char *argv[])
 		}
 #endif
 
+		if (strcmp(argv[1],"bootflag") == 0) {
+			ulong flag = simple_strtoul(argv[2], NULL, 16);
+			if (0x5A5A5A5A != flag && 0xA5A5A5A5 != flag) {
```
Comment (truongn1) [inline common/cmd_nvedit.c]: change to use macro value

File: common/cmd_bootm.c (hunk 1 above)
Comment (ghost) [reply common/cmd_bootm.c]: @truongn1 You are absolutely right, but I didn't have enough time for a perfect design. It was too rushed

File: common/cmd_nvedit.c (hunk 2 above)
Comment (ghost) [reply common/cmd_nvedit.c:211]: @truongn1 You are definitely right. It's a rushed design

File: net/httpd.c (hunk 3 above)
Comment (ghost) [reply net/httpd.c]: Of course. This is just a rushed design for quick testing

File: common/cmd_nvedit.c (hunk 2 above)
Comment (truongn1) [inline common/cmd_nvedit.c:211]: Should we add `sys_bootflag = BOOT_FROM_A` here? As a fall back value?

Comment (ghost) [reply common/cmd_nvedit.c:211]: please check the latest commit on this PR

**Review by truongn1 — APPROVED** [review]

**Pull request merged**